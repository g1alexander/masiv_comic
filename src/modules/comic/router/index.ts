import type { RouteRecordRaw } from 'vue-router'

const router: RouteRecordRaw = {
  path: '/',
  name: 'comic',
  component: () => import('../layout/ComicLayout.vue'),
  children: [
    {
      path: '',
      name: 'comic-home',
      component: () => import('../views/HomeView.vue')
    }
  ]
}

export default router
