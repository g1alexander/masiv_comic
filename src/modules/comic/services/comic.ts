import { xkcdApi } from '@/api/xkcdApi'
import type { Comic, ResponseComicAdapter } from '.'
import { comicAdapter } from '../helpers/comicAdapter'

export async function comic(page: number): Promise<ResponseComicAdapter> {
  try {
    const { data } = await xkcdApi().get<Comic>(`/comic?num=${page}`)

    const response = comicAdapter(data)

    return response
  } catch (error) {
    return {
      alt: '',
      img: '',
      num: 0,
      title: '',
      transcript: '',
      stars: 0
    }
  }
}
