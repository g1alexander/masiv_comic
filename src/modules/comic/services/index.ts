export * from './comic'
export * from './interfaces/Comic'
