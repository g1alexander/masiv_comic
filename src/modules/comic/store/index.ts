import { defineStore } from 'pinia'
import { getExtractStore } from '@/shared/helpers/getExtractStore'
import { useState } from './state'
import { useGetters } from './getters'
import { useActions } from './actions'

export const useComicStore = defineStore('comic', () => ({
  ...getExtractStore(useState()),
  ...getExtractStore(useGetters()),
  ...getExtractStore(useActions())
}))
