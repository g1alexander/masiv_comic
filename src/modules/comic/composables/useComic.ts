import { onMounted, watch } from 'vue'
import { useComicStore } from '../store'

export function useComic() {
  const comic = useComicStore()

  onMounted(async () => {
    await comic.setComic()
  })

  watch(
    () => comic.getPage,
    async () => await comic.setComic()
  )

  const setStar = (star: number) => comic.setStar(star)

  const starClass = (star: number) => {
    return {
      'star-active': star <= comic.getComic.stars,
      'star-disabled': star > comic.getComic.stars
    }
  }
  return {
    comic,
    setStar,
    starClass
  }
}
