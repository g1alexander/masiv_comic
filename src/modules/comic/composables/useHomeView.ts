import { defineAsyncComponent, onMounted } from 'vue'
import { useComicStore } from '../store'
import loader from '@/assets/icons/loader.svg'

export function useHomeView() {
  const comic = useComicStore()

  onMounted(async () => {
    await comic.setComic()
  })

  return {
    loader,
    comic,
    HomeInfo: defineAsyncComponent(() => import('../components/HomeInfo.vue')),
    HomeActions: defineAsyncComponent(() => import('../components/HomeActions.vue'))
  }
}
