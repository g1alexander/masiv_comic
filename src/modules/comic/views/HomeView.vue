<script lang="ts" setup>
import { useComic } from '../composables'

const { loader, comic, setStar, starClass } = useComic()
</script>

<template>
  <main class="home">
    <template v-if="comic.getIsLoading">
      <img :src="loader" alt="loading..." width="180" />
    </template>

    <section class="home-section" v-else>
      <h1 class="home-section__h1">{{ comic.getComic.title.toUpperCase() }}</h1>
      <img class="home-section__img" :src="comic.getComic.img" :alt="comic.getComic.alt" />

      <div class="home-section__stars">
        <h3>Puntuacion:</h3>
        <p>
          <template v-for="star in 5" :key="star">
            <i
              @click="setStar(star)"
              class="home-section__stars__icon"
              :class="starClass(star)"
            ></i>
          </template>
        </p>
      </div>

      <article class="home-section__btns">
        <button v-if="comic.getPage > 1" @click="comic.setPage(comic.getPage - 1)">prev</button>
        <button @click="comic.setPage(comic.getPage + 1)">next</button>
      </article>
    </section>
  </main>
</template>
