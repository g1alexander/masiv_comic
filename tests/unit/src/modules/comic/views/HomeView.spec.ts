import { describe, expect, test, vi } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import HomeView from '@/modules/comic/views/HomeView.vue'
import { useHomeView } from '@/modules/comic/composables'
import { dataComponent } from '@tu/mocks/data'

describe('component HomeView.vue', () => {
  vi.mock('@/modules/comic/composables/useHomeView', () => {
    const useHomeView = vi.fn(() => dataComponent)

    return { useHomeView }
  })

  test('to match snapshot', () => {
    const wrapper = shallowMount(HomeView)

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should show loader if loading status is true', () => {
    const wrapper = shallowMount(HomeView)
    expect(wrapper.find('img').attributes()).toEqual({
      alt: 'loading...',
      src: '/src/assets/icons/loader.svg',
      width: '180'
    })
  })

  test('should show comic if loading status is false', () => {
    const data = { ...dataComponent, comic: { getIsLoading: false } as any }

    vi.mocked(useHomeView).mockReturnValue(data)

    const wrapper = shallowMount(HomeView)

    const sectionHome = wrapper.find('.home-section')
    const homeInfo = wrapper.find('home-info-stub')
    const homeActions = wrapper.find('home-actions-stub')

    expect(sectionHome.exists()).toBeTruthy()
    expect(homeInfo.exists()).toBeTruthy()
    expect(homeActions.exists()).toBeTruthy()
  })
})
