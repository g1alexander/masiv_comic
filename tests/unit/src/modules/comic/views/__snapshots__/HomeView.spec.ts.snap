// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`component HomeView.vue > to match snapshot 1`] = `"<main class=\\"home\\"><img src=\\"/src/assets/icons/loader.svg\\" alt=\\"loading...\\" width=\\"180\\"></main>"`;
